apps/pipeline-visualization/src/mermaid_utils.ts

```diff
@@ -0,0 +1,156 @@
+import { 
+    JayveeModel, 
+    JayveeServices, 
+    createJayveeServices, 
+    PipelineDefinition, 
+    getBlocksInTopologicalSorting,
+    collectStartingBlocks,
+    collectChildren, 
+    BlockDefinition
+  } from '@jvalue/jayvee-language-server';
+import { integer } from 'vscode-languageserver-protocol';
+
+import { 
+  diagramType,
+  diagramDirection, 
+  subgraphDirection,
+  subgraphColor,
+  //showComposite,
+  //showProperties,
+  properties,
+  font,
+  fontSize
+  } from './mermaid_params';
+
+
+  export interface MermaidOptions {
+    composite: boolean;
+    properties: boolean;
+  }
+
+function processPipeline(pipeline: PipelineDefinition): string{
+  let listofPipes: Array<string[]> = []
+  const process_pipe = (pipe: string[], block:BlockDefinition) => {
+    pipe.push(block.name)
+    let children = collectChildren(block);
+    if (children.length == 1) {
+      process_pipe(pipe, children[0]!);
+    } else if (children.length > 1){
+      listofPipes.push(pipe);
+      children.forEach((child) => {process_pipe([block.name], child)})
+    } else {
+      listofPipes.push(pipe)
+    }
+  }
+  let startingBlocks = collectStartingBlocks(pipeline);
+  startingBlocks.forEach((startingBlock) => process_pipe([], startingBlock))
+  let result = listofPipes
+  .map((pipeline) => pipeline.join("-->"))
+  .join("\n")
+  return result;
+}
+
+export function createMermaidPipeline(pipeline: PipelineDefinition, mermaidOptions: MermaidOptions){
+  /* const myToString = (block: BlockDefinition, index = 0): string => {
+    const blockTypeName = block.type.ref?.name;
+    const blockString = `${index}["\`${block.name} \n (${blockTypeName})\`"]`;
+    const childString = collectChildren(block)
+      .map((child) => myToString(child, index + 1))
+      .join('-->');
+    return blockString + '-->' + childString;
+  };
+ */
+  let name: string = "subgraph " + pipeline.name;
+  let direction: string = "direction " + subgraphDirection
+  let listofPipes: Array<string[]> = [];
+  let listofBocks: Array<string> = [];
+  let composites: string = "";
+  const process_pipe = (pipe: string[], block:BlockDefinition) => {
+    if (block.type.ref?.$type == "CompositeBlocktypeDefinition" && mermaidOptions.composite){
```

Comment:
Try working with the AST and not on the structure of the objects to make the code more robust.

Examples of useful imports from `@jvalue/jayvee-language-server`:
```
isChainedPipeDefinition
isBlocktypePipeline
isCompositeBlocktypeDefinition

BlockTypeWrapper

isTextLiteral
```